COMPAT: reading generic PyTables Table format fails with sub-selection

jgehrcke opened the pull request:
- [x] closes #11188
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

jgehrcke committed:
TST: add TestReadPyTablesHDF5 test scaffold

simonjayhawkins commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5167,3 +5168,75 @@ def test_dst_transitions(self):
+    (not written by pandas).
+    """
+
+    def _create_simple_hdf5_file_with_pytables(self):
```
is this necessary or could you just add a file in pandas\tests\io\data

jgehrcke replied:
> is this necessary or could you just add a file in pandas\tests\io\data

Oh, that's some early feedback. Thank you!

I think a proper test for the fix for #11188 should confirm that the minimal working example for reproducing the problem does not fail anymore.

I also think this is more readable (when the test itself shows how the data file was created, in contrast to committing a binary file of which it's not 100 % clear and discoverable of how it has been created).

Also, I'd like to think of this (using the `PyTables` API to create an HDF5 file) as an important aspect of this test: it effectively confirms a common way to use `PyTables` and `pandas` in tandem; `PyTables` for writing an HDF5 file, and `pandas` for reading it. These tests are not so much about the contents in the file, but about that a specific way to create the file matches a specific way to parse the file.

Well, let's see about the "fix", looking forward to your feedback there. I think this will be more controversial.

jgehrcke committed:
BUG: this fixes #11188

pep8speaks commented:
Hello @jgehrcke! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/io/pytables.py`](https://github.com/pandas-dev/pandas/blob/cce0e576a8bb9b5080d8a1ced4aff9baff2f14bd/pandas/io/pytables.py):

> [Line 1828:9](https://github.com/pandas-dev/pandas/blob/cce0e576a8bb9b5080d8a1ced4aff9baff2f14bd/pandas/io/pytables.py#L1828): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

jgehrcke committed:
DOC: update whatsnew for #11188

jgehrcke commented:
Funny `codecov` bot :-). Well, let's wait for this round of CI to report back.

jgehrcke commented:
@gfyoung @simonjayhawkins @jreback I would appreciate your feedback, especially about the "fix" itself for starters, and not so much about the tests. The workaround is not necessarily nice. All tests pass, but could it be that this approach (introducing `table._nrows_to_read = stop - start` dynamically, and using it dynamically instead of `table.nrows`) has negative side effects?

gfyoung commented on pandas/io/pytables.py:
```diff
@@ -1459,6 +1459,11 @@ def __init__(self, store, s, func, where, nrows, start=None, stop=None,
+            # Piggy-back normalized `nrows` (considering start and stop) onto
+            # PyTables tables object so that the GenericIndexCol constructor
+            # knows how large the index should be.
+            self.s.table._nrows_to_read = stop - start
```
Why not make these class attributes of `self` ?

jgehrcke replied:
Huhu!

> start and stop - why not make these instance attributes of self ?

Hm, what do you have in mind about doing with them after having set them as instance attributes of `self`? In the three lines below we're doing exactly that (we set `self.start` and self.stop`; did you mean something else?)

jgehrcke replied:
> Or even just compute nrows as an attribute?

Interesting. I thought about this a bit but I am not sure what you mean. Where; on the `PyTables` table object?

That would be quite invasive. A `PyTables` table object might use its `nrows` property internally. 

Adding a new private attribute to an object (guaranteed to not mess with the object's intrinsics) is not beautiful, but a little less invasive than overwriting an object property which might mess with that object's internal consistency.

gfyoung replied:
@jgehrcke : I'm talking about attaching a new attribute to the `TableIterator` instance itself.

Attaching attributes to the `TableIterator`'s `s.table` is riskier.

jgehrcke replied:
> @jgehrcke : I'm talking about attaching a new attribute to the TableIterator instance itself.

Gotcha! I'll look into that. Thanks for the feedback, much appreciated. 

> Attaching attributes to the TableIterator's s.table is riskier.

Ack!

jgehrcke replied:
There is however no obvious way for me to access the `TableIterator` instance from the `GenericIndexCol` instance. The latter inherits from `IndexCol` which gets its `.table` attribute set through a `set_table()` method, but as far as I can see there is no existing obvious relationship to reach back to the `TableIterator`... I think I had the problem before which is why I chose the table object to piggy-back the piece of information on.

jgehrcke replied:
@gfyoung so this is the context where we call into the `convert()` method of the `GenericIndexCol`:

https://github.com/pandas-dev/pandas/blob/a60888ce4ce9e106537fb410688b66baa109edc3/pandas/io/pytables.py#L3437

There are certainly other ways to pass along the `_nrows_to_read` concept but I didn't see a cleaner way yet, specifically I didn't see a way that does not add additional scaffold, passing arguments around etc.

Did you have something very specific in mind that I simply didn't see?

gfyoung replied:
Interesting...we might have to then propagate an `nrows` parameter throughout the `IndexCol` instantiations.

@jreback : Thoughts?

jgehrcke replied:
> Attaching a parameter to an object we have no information about is definitely riskier and hackier (and is therefore a practice I would discourage) compared to passing arguments around.

I do agree.

jgehrcke committed:
CLN: do not piggyback idx size on table obj

jgehrcke commented:
@gfyoung @jreback after following the quite entangled code path I think I found a cleaner solution.

Key insight was the following:

In `HDFStore.select()` we define a callback `func`:
```python
        # function to call on iteration
        def func(_start, _stop, _where):
            return s.read(start=_start, stop=_stop,
                          where=_where,
                          columns=columns)

        # create the iterator
        it = TableIterator(self, s, func, where=where, nrows=s.nrows,
                           start=start, stop=stop, iterator=iterator,
                           chunksize=chunksize, auto_close=auto_close)

        return it.get_result()
```

In the problematic code path `s.read()` actually translates to `Table.read_axes()` which is prepared to receive arbitrary keyword arguments:

```python
    def read_axes(self, where, **kwargs):
        """create and return the axes sniffed from the table: return boolean
        for success
        """
        ...
```

As of the callback being `s.read(start=_start, stop=_stop, ....)` we know that the `start` and `stop` parameters are available as keyword arguments within `Table.read_axes()`.

`read_axes()` calls into `GenericIndexCol.convert()` and so this is a decent opportunity to pass the `start` and `stop` parameters directly to `convert()`. I have added a corresponding commit. Please have another look, thanks!

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -656,6 +656,7 @@ I/O
 - Bug in :func:`read_csv` not properly interpreting the UTF8 encoded filenames on Windows on Python 3.6+ (:issue:`15086`)
 - Improved performance in :meth:`pandas.read_stata` and :class:`pandas.io.stata.StataReader` when converting columns that have missing values (:issue:`25772`)
 - Bug in :meth:`DataFrame.to_html` where header numbers would ignore display options when rounding (:issue:`17280`)
+- Bug in :func:`read_hdf` where the `start` and `stop` arguments would raise a ``ValueError`` indicating an index size mismatch (:issue:`11188`)
```
this is not very informative of what you are actually addressing, can you re-word

jgehrcke replied:
commit upcoming

jreback commented on pandas/io/pytables.py:
```diff
@@ -1627,7 +1627,7 @@ def infer(self, handler):
 
-    def convert(self, values, nan_rep, encoding, errors):
+    def convert(self, values, nan_rep, encoding, errors, **kwargs):
         """ set the values from this selection: take = take ownership """
```
can you add a doc-string here

jreback replied:
types are a bonus

jreback replied:
I would just add ``start=None`` and ``stop=None`` instead of kwargs

jgehrcke replied:
> I would just add start=None and stop=None instead of kwargs

done, commit upcoming

> can you add a doc-string here

I would love to add one! I have difficulties understanding the general purpose of the `convert()` method though and documenting `values`, `nan_rep`, `encoding`, and `errors` is pretty challenging given my lack of understanding. As far as I see the `start` and `stop` arguments are not really meaningful arguments here (in `IndexCol.convert()`), so I am also a bit helpless with documenting them! :-) How would you document `start` and `stop` here?

jreback commented on pandas/io/pytables.py:
```diff
@@ -1816,10 +1816,14 @@ class GenericIndexCol(IndexCol):
 
-    def convert(self, values, nan_rep, encoding, errors):
+    def convert(self, values, nan_rep, encoding, errors, **kwargs):
         """ set the values from this selection: take = take ownership """
```
doc-string here

jgehrcke replied:
:+1: adding a docstring explaining the `start` and stop` parameters, commit upcoming

jreback commented on pandas/io/pytables.py:
```diff
@@ -2162,7 +2166,7 @@ def validate_attr(self, append):
 
-    def convert(self, values, nan_rep, encoding, errors):
+    def convert(self, values, nan_rep, encoding, errors, **kwargs):
         """set the data from this selection (and convert to the correct dtype
```
same

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5167,3 +5168,75 @@ def test_dst_transitions(self):
+        ]
+
+        # This returns a path and does not open the file.
+        tmpfilepath = create_tempfile(self.path)
```
so this should just be a fixture

jgehrcke replied:
:+1: 

I now turned the entire thing into a module-scoped fixture, providing the path to the HDF5 file and details about its contents.

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5167,3 +5168,75 @@ def test_dst_transitions(self):
+
+    def _compare(self, df, samples):
+        """Compare the reference `samples` with the contents in DataFrame `df`.
+        """
```
don't write a custom comparator, use assert_frame_equal

jgehrcke replied:
Thanks.

Interesting:

```
E       AssertionError: Attributes are different
E       
E       Attribute "dtype" are different
E       [left]:  uint32
E       [right]: int64
```

But that's probably expected, right? So, we expect the values in the `UInt32Col` in the PyTables file to end up being of type int64, right?

jgehrcke replied:
Okay, thanks for the review, I have committed and pushed changes and would appreciate another round of feedback.

:rocket:

jgehrcke committed:
DOC: improve changelog (squash later)

jgehrcke committed:
squash: remove kwargs, add a bit of docstring

jgehrcke committed:
squash: rework tests (use fixture, assert_frame_equal)

jgehrcke commented on pandas/io/pytables.py:
```diff
@@ -1816,10 +1816,24 @@ class GenericIndexCol(IndexCol):
+            Table row number: the start of the sub-selection.
+        stop : int, optional
+            Table row number: the end of the sub-selection. Values larger than
+            the underlying table's row count are normalized to that.
```
At least here I understand the meaning of `start` and `stop` and have tried to document it.

jreback replied:
right, can you add all the parameters. See if you can put something down for them, some doc-string is better than none.

jgehrcke replied:
:+1: 

Adding statements for `values`, `nan_rep`, `encoding`, and `errors`. 

The biggest challenge for me still is that I do not understand the general purpose of the convert() method and its arguments.

jgehrcke commented:
Will polish (squash commits, address line length issues) after another round of comments.

jgehrcke committed:
squash: address line length errors

jgehrcke committed:
squash: pytables.py: remove one assignment

jreback commented on pandas/io/pytables.py:
```diff
@@ -1816,10 +1817,25 @@ class GenericIndexCol(IndexCol):
+        """
+
+        start = start if start is not None else 0
+        stop = min(stop, self.table.nrows) \
```
use perens rather  than a line-continuation

jgehrcke replied:
commit upcoming

jgehrcke replied:
done

jreback commented on pandas/io/pytables.py:
```diff
@@ -3434,8 +3451,10 @@ def read_axes(self, where, **kwargs):
+            # `store.select()`. If set they determine the index size.
             a.convert(values, nan_rep=self.nan_rep, encoding=self.encoding,
-                      errors=self.errors)
+                      errors=self.errors, **kwargs)
```
where is this kwargs coming from? do you mean to pass ing start/stop?

jgehrcke replied:
> where is this kwargs coming from

I have tried to answer this question with the code comment added right above that line, and all I know is that kwargs _may_ contain `start` and `stop`. I could not infer any guarantees.

Following the kwargs flow(s) is a challenging part of reading `io/pytables.py`. It's pretty non-obvious.


There are two calls into that function that pass `**kwargs` where it's non-obvious where it's coming from:

https://github.com/pandas-dev/pandas/blob/d47947a79099d650b4bb981a38ce09881ec6f43d/pandas/io/pytables.py#L3888

https://github.com/pandas-dev/pandas/blob/d47947a79099d650b4bb981a38ce09881ec6f43d/pandas/io/pytables.py#L4143

For following just the latter flow I am now looking for `read()` being called and this is where things get tricky...
```
$ grep -nr 'read(' pandas/io/pytables.py | grep -v 'def ' | grep -v validate_read
721:            return s.read(start=_start, stop=_stop,
837:            objs = [t.read(where=_where, columns=columns, start=_start,
1421:        return s.read(**kwargs)
2934:            sdict[c] = s.read()
4241:        s = super().read(columns=columns, **kwargs)
4342:        df = super().read(**kwargs)
4740:        return self.table.table.read(start=self.start, stop=self.stop)
```
From here it looks like there is more kwargs to follow. I know one definite call path that comes along line 837 and there we _know_ that `start` and `stop` are being passed explicitly (that's the code path coming from `read_hdf()`). But I am not sure about other possible code paths.

Relying on the test suite to not break, but also relying on _not_ relying on `start` and `stop` being passed explicitly.

I am now doing this instead:
```
-                      errors=self.errors, **kwargs)
+                      errors=self.errors, start=kwargs.get('start'),
+                      stop=kwargs.get('stop'))
```
So that things still don't break when `start` or `stop` are not in `kwargs`. Sound good?

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5167,3 +5168,71 @@ def test_dst_transitions(self):
                 assert_frame_equal(result, df)
+
+
+@pytest.fixture(scope='module')
```
this is not necessary, function scoped is fine

jgehrcke replied:
:+1: commit upcoming

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5167,3 +5168,71 @@ def test_dst_transitions(self):
+        {'c0': t0 + 2, 'c1': 'ccccc', 'c2': 10**5},
+        {'c0': t0 + 3, 'c1': 'ddddd', 'c2': 4294967295},
+    ]
+
```
use ensure_clean_path here

jgehrcke replied:
Done, commit upcoming.

Food for thought, I don't expect answers for this pull request here:

- If a test fails, don't we want to retain the corresponding HDF5 file in the file system to enable later inspection? If the answer is "yes" then we should not use a context manager for file deletion, but do it in the fixture teardown only upon test success.

- What do we think about using a pytest-provided fixture for temporary file management?

```python
@pytest.fixture()
def pytables_hdf5_file(tmp_path):
    [...]
```

In that case the file path would contain the test name, e.g. `tmp_path = PosixPath('PYTEST_TMPDIR/test_create_file0')` as documented here: https://docs.pytest.org/en/latest/tmpdir.html

jreback replied:
ensure_clean_path already does this, not really necessary to do anything else

jgehrcke replied:
> ensure_clean_path already does this

The way I read the code it will _always_ remove the file, also upon test failure.

Anyway :).

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5167,3 +5168,71 @@ def test_dst_transitions(self):
+    return tmpfilepath, objectname, pd.DataFrame(testsamples)
+
+
+class TestReadPyTablesHDF5(Base):
```
do you *need* to inherit from Base?

jgehrcke replied:
> do you need to inherit from Base?

I do not really know. Not sure if `tm.reset_testing_mode()`  is important at all in my context here:
```\class Base:

    @classmethod
    def setup_class(cls):

        # Pytables 3.0.0 deprecates lots of things
        tm.reset_testing_mode()

    @classmethod
    def teardown_class(cls):

        # Pytables 3.0.0 deprecates lots of things
        tm.set_testing_mode()
```

Given that you doubt that I need it I have added a commit and don't inherit from `Base` anymore. The tests still pass on my system.

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5167,3 +5168,71 @@ def test_dst_transitions(self):
+
+    def test_read_complete(self, pytables_hdf5_file):
+        path, objname, expected_df = pytables_hdf5_file
+        assert_frame_equal(pd.read_hdf(path, key=objname), expected_df)
```
write the asserts like

result = 
expected = 
assert_frame_equal(result, expected)

jgehrcke replied:
done, thanks

jgehrcke committed:
squash: tests: result/expected, Base, fixture scope

jgehrcke committed:
CLN: s/existing/exiting

jgehrcke committed:
TST: use ensure_clean_path context manager

jgehrcke committed:
squash: do not use line cont backslash

jgehrcke committed:
squash: pass start/stop explicitly into a.convert()

jgehrcke committed:
DOC: add some docstring

jgehrcke committed:
squash: flake8 error

jgehrcke commented:
@jreback I tried to address all points in your last round of feedback. I feel like we're getting closer. Would very much appreciate another look, thanks for your time :clock10: !

jreback commented on pandas/io/pytables.py:
```diff
@@ -1816,10 +1817,33 @@ class GenericIndexCol(IndexCol):
+        ----------
+
+        values :
+            Expected to be passed but ignored in this implementation.
```
can you write that these are, e.g.

values: np.ndarray
 nan_rep : str
encoding: str
errors : str (I think)

jreback commented on pandas/io/pytables.py:
```diff
@@ -1816,10 +1817,33 @@ class GenericIndexCol(IndexCol):
+        values :
+            Expected to be passed but ignored in this implementation.
+        nan_rep :
+            Expected to be passed but ignored in this implementation.
```
don't need these comments

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5,6 +5,7 @@
 from io import BytesIO
 import os
 import tempfile
+import time
```
what are you using this for?

jgehrcke replied:
Used this for dynamically generating unix time stamps (`time.time()`) before writing them to the HDF5 file, but made this static now.

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5167,3 +5168,70 @@ def test_dst_transitions(self):
                 assert_frame_equal(result, df)
+
+
+@pytest.fixture()
```
you don't need the ()

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5167,3 +5168,70 @@ def test_dst_transitions(self):
                 store.append('df', df)
                 result = store.select('df')
                 assert_frame_equal(result, df)
+
```
can you move these to a separate (new) file: test_pytables_compat.py (even better to move to

pandas/tests/io/pytables/test_compat.py (you will need to add an __init__.py in the subdir)

we are starting to split the test_pytables up a bit

jgehrcke replied:
Doing so.

> (you will need to add an init.py in the subdir)

Why?

It does not need to be a Python package for pytest to discover tests in the directory:
```
$ pytest --collect-only pandas/tests
[...]
<Module pandas/tests/io/pytables/test_compat.py>
  <Class TestReadPyTablesHDF5>
      <Function test_read_complete>
      <Function test_read_with_start>
      <Function test_read_with_stop>
      <Function test_read_with_startstop>
[...]
```
Any other reason?

jgehrcke committed:
squash: change docstrings one more time

jgehrcke committed:
squash: tests: cleanup

jgehrcke committed:
TST: move tests to io/pytables/test_compat.py

jgehrcke committed:
TST: use pytest fixture instead of ensure_clean_path

jgehrcke commented:
@jreback thanks for your feedback once again.

I have
- rebased on upstream/master (clean)
- added four new commits to address your last comments

In particular, I have now moved the new tests to `pandas/tests/io/pytables/test_compat.py` as suggested. While doing so I have not created a  `pandas/tests/io/pytables/__init__.py` file because I don't see the reason yet (it's not for test discovery, but is it for linters?).

I have also **not** used `ensure_clean_path` (because then I would have needed to refactor it out of where it is right now). Instead I have used the pytest `tmp_path` fixture which I think provides everything we need with less code: it guarantees to not have file system collision between tests, it retains files for later inspection (as opposed to the `ensure_clean_path` context manager) and removes them in a rolling fashion so that storage needs do not grow indefinitely with individual invocation. For instance, this is after invoking pytest 11 times (files from the last three invocations are retained, the rest got rotated away):

```
$ tree
.
├── pytest-10
│   ├── test_read_complete0
│   │   └── written_with_pytables.h5
│   ├── test_read_completecurrent -> /tmp/pytest-of-jp/pytest-10/test_read_complete0
│   ├── test_read_with_start0
│   │   └── written_with_pytables.h5
│   ├── test_read_with_startcurrent -> /tmp/pytest-of-jp/pytest-10/test_read_with_start0
│   ├── test_read_with_startstop0
│   │   └── written_with_pytables.h5
│   ├── test_read_with_startstopcurrent -> /tmp/pytest-of-jp/pytest-10/test_read_with_startstop0
│   ├── test_read_with_stop0
│   │   └── written_with_pytables.h5
│   └── test_read_with_stopcurrent -> /tmp/pytest-of-jp/pytest-10/test_read_with_stop0
├── pytest-8
│   ├── test_read_complete0
│   │   └── written_with_pytables.h5
│   ├── test_read_completecurrent -> /tmp/pytest-of-jp/pytest-8/test_read_complete0
│   ├── test_read_with_start0
│   │   └── written_with_pytables.h5
│   ├── test_read_with_startcurrent -> /tmp/pytest-of-jp/pytest-8/test_read_with_start0
│   ├── test_read_with_startstop0
│   │   └── written_with_pytables.h5
│   ├── test_read_with_startstopcurrent -> /tmp/pytest-of-jp/pytest-8/test_read_with_startstop0
│   ├── test_read_with_stop0
│   │   └── written_with_pytables.h5
│   └── test_read_with_stopcurrent -> /tmp/pytest-of-jp/pytest-8/test_read_with_stop0
├── pytest-9
│   ├── test_read_complete0
│   │   └── written_with_pytables.h5
│   ├── test_read_completecurrent -> /tmp/pytest-of-jp/pytest-9/test_read_complete0
│   ├── test_read_with_start0
│   │   └── written_with_pytables.h5
│   ├── test_read_with_startcurrent -> /tmp/pytest-of-jp/pytest-9/test_read_with_start0
│   ├── test_read_with_startstop0
│   │   └── written_with_pytables.h5
│   ├── test_read_with_startstopcurrent -> /tmp/pytest-of-jp/pytest-9/test_read_with_startstop0
│   ├── test_read_with_stop0
│   │   └── written_with_pytables.h5
│   └── test_read_with_stopcurrent -> /tmp/pytest-of-jp/pytest-9/test_read_with_stop0
└── pytest-current -> /tmp/pytest-of-jp/pytest-10
```

jgehrcke committed:
squash: address linter errors

jgehrcke committed:
squash: pass path as text

jgehrcke committed:
squash: tests: improve code comment

jreback commented on pandas/tests/io/pytables/test_compat.py:
```diff
@@ -0,0 +1,84 @@
+
+    objname = 'pandas_test_timeseries'
+
+    # The `tmp_path` fixture provides a temporary directory unique to the
```
can you use ensure_clean_path; we don't use the pytest fixture at this time, though you can open an issue to change (many occurrences of this I suppose).

jreback reviewed (changes requested):
lgtm, but need to clean up the test files, ping on green.

jreback commented on pandas/tests/io/pytables/test_compat.py:
```diff
@@ -0,0 +1,84 @@
+def pytables_hdf5_file(tmp_path):
+    """Use PyTables to create a simple HDF5 file.
+
+    There is no need for temporary file cleanup: pytest's `tmp_path` fixture
```
remove this comment as well. see my comment below. we don't want to introduce a different idiom for testing (but as I said you can open an issue for discussion).

jgehrcke replied:
> you can open an issue for discussion

 https://github.com/pandas-dev/pandas/issues/26984 :)

jgehrcke committed:
squash: use ensure_clean_path again

jreback commented on pandas/tests/io/pytables/test_compat.py:
```diff
@@ -0,0 +1,76 @@
+
+
+@pytest.fixture
+def pytables_hdf5_file(tmp_path):
```
I don't think you need this fixture anymore

jgehrcke replied:
> I don't think you need this fixture anymore 

Thanks man. I hope one of the linters would have told me! :-)

jreback replied:
yeah i don’t think we have that check in linting :)

jreback commented on pandas/tests/io/pytables/test_compat.py:
```diff
@@ -0,0 +1,76 @@
+
+class TestReadPyTablesHDF5:
+    """
+    A group of tests which covers reading HDF5 files written by plain PyTables
```
can you add the issue number here in the comment

jgehrcke committed:
squash: tests: cleanup

jgehrcke committed:
squash: tests: fix isort error

jgehrcke commented:
@jreback it's green now. Wanna have another look before I squash some commits?

jreback commented:
thanks @jgehrcke nice patch. We squash on merge anyhow, so no need!

jreback commented:
If interested would like to move the existing test_pytables test files to pandas/test/io/pytables (just a simple move). We want to split the main file, but that will be separate.

jgehrcke commented:
> thanks @jgehrcke nice patch

Thank you for the nice collaboration!

jgehrcke commented:
> If interested would like to move the existing test_pytables test files (test_pytables.py and test_pytables_missing.py) to pandas/test/io/pytables (just a simple move). 

Sure, PR upcoming.
